usersync/syncersbuilder.go

```diff
@@ -98,28 +101,88 @@ func chooseSyncerConfig(biddersSyncerConfig []namedSyncerConfig) (namedSyncerCon
+	)
+
 	for _, bidder := range biddersSyncerConfig {
 		bidderNames = append(bidderNames, bidder.name)
 		if bidder.cfg.IFrame != nil || bidder.cfg.Redirect != nil {
-			bidderNamesWithEndpoints = append(bidderNamesWithEndpoints, bidder.name)
+			if len(bidder.bidderInfo.AliasOf) > 0 {
+				aliasBidderNames = append(aliasBidderNames, bidder.name)
+				parentBidderNameMap[bidder.bidderInfo.AliasOf] = struct{}{}
+			} else {
+				nonAliasBidderNames = append(nonAliasBidderNames, bidder.name)
+			}
 			syncerConfig = bidder
 		}
 	}
 
-	if len(bidderNamesWithEndpoints) == 0 {
+	// check if bidders have same syncer key but no endpoints
+	if len(nonAliasBidderNames)+len(aliasBidderNames) == 0 {
 		sort.Strings(bidderNames)
 		bidders := strings.Join(bidderNames, ", ")
 		return namedSyncerConfig{}, fmt.Errorf("bidders %s share the same syncer key, but none define endpoints (iframe and/or redirect)", bidders)
 	}
 
-	if len(bidderNamesWithEndpoints) > 1 {
-		sort.Strings(bidderNamesWithEndpoints)
-		bidders := strings.Join(bidderNamesWithEndpoints, ", ")
+	// check if non-alias bidders have same syncer key
+	if len(nonAliasBidderNames) > 1 {
+		sort.Strings(nonAliasBidderNames)
+		bidders := strings.Join(nonAliasBidderNames, ", ")
 		return namedSyncerConfig{}, fmt.Errorf("bidders %s define endpoints (iframe and/or redirect) for the same syncer key, but only one bidder is permitted to define endpoints", bidders)
 	}
 
+	// check if alias bidders of different parent have same syncer key
+	if len(parentBidderNameMap) > 1 {
+		sort.Strings(aliasBidderNames)
+		return namedSyncerConfig{}, fmt.Errorf("alias bidders %s of different parents defines endpoints (iframe and/or redirect) for the same syncer key, but only one bidder is permitted to define endpoints", strings.Join(aliasBidderNames, ", "))
+	}
+
+	// check if aliases of same parent and non-alias bidder have same syncer key
+	if len(parentBidderNameMap) != 0 {
+		if _, ok := parentBidderNameMap[nonAliasBidderNames[0]]; !ok {
+			sort.Strings(aliasBidderNames)
+			return namedSyncerConfig{}, fmt.Errorf("alias bidders %s and non-alias bidder %s defines endpoints (iframe and/or redirect) for the same syncer key, but only one bidder is permitted to define endpoints", strings.Join(aliasBidderNames, ", "), nonAliasBidderNames[0])
+		}
+	}
+
 	return syncerConfig, nil
 }
+
+func getSyncerKey(biddersWithSyncerCfg map[string]config.BidderInfo, bidderName string, bidderInfo config.BidderInfo) (string, error) {
+	if bidderInfo.Syncer == nil {
+		return "", fmt.Errorf("found no syncer config for alias bidder %s", bidderName)
+	}
+
+	var (
+		bidderSyncerCfg       = bidderInfo.Syncer
+		isAlias               = len(bidderInfo.AliasOf) > 0
+		hasInheritedSyncerCfg = false
+		parentBidderInfo      config.BidderInfo
+		parentBidderName      string
+	)
+
+	if isAlias {
+		parentBidderName = bidderInfo.AliasOf
+		var parentHasSyncerCfg bool
+		parentBidderInfo, parentHasSyncerCfg = biddersWithSyncerCfg[parentBidderName]
```

Comment:
I think this line can be conslidated into just

```
parentBidderInfo, parentHasSyncerCfg := biddersWithSyncerCfg[parentBidderName]
```

Reply:
Map yields bidderName to bidderInfo. so - `parentHasSyncerCfg ` value should not be derived from just checking that map. Also we should not ignore if parent bidder is not found in map. However that case is highly unlikely. 
so - 
```
if parentBidderInfo, ok := bidderInfos[parentBidderName]; !ok {
    //return err
}

Reply:
https://github.com/prebid/prebid-server/pull/3082#discussion_r1319178458
> I think this line can be conslidated into just
> 
> ```
> parentBidderInfo, parentHasSyncerCfg := biddersWithSyncerCfg[parentBidderName]
> ```

@AlexBVolcy , we can consildate to above. Because `parentBidderInfo` is also referred outside of `if ` block and scope of `parentHasSyncerCfg` is within the if block. So both `parentBidderInfo & parentHasSyncerCfg ` are declared and initialized separately. Go won't allow this by using `:=`

```

L161 	var (
	        ...
L165		parentBidderInfo      config.BidderInfo <--- parentBidderInfo declaration
	        ...
L167 	)

        ...

L169 	if isAlias {
L170    	parentBidderName = bidderInfo.AliasOf 
L171    	var parentHasSyncerCfg bool <--- parentHasSyncerCfg declaration
L172    	parentBidderInfo, parentHasSyncerCfg = biddersWithSyncerCfg[parentBidderName] <--- parentBidderInfo initialization
L173	    hasInheritedSyncerCfg = parentHasSyncerCfg && parentBidderInfo.Syncer == bidderSyncerCfg  <---- parentHasSyncerCfg usage limited to if block
L174  	} 
        ...
L183    if isAlias && !hasInheritedSyncerCfg && bidderSyncerCfg.Key == parentBidderInfo.Syncer.Key { <--- parentBidderInfo refers outside if block
L184		return "", fmt.Errorf("syncer key of alias bidder %s is same as the syncer key for its parent bidder %s", bidderName, parentBidderName)
L185	}
```



---------------------------------------------------------------------------------------------------------------------------------------




https://github.com/prebid/prebid-server/pull/3082#discussion_r1319409218
> Map yields bidderName to bidderInfo. so - `parentHasSyncerCfg ` value should not be derived from just checking that map. Also we should not ignore if parent bidder is not found in map. However that case is highly unlikely. so -
> 
> ```
> if parentBidderInfo, ok := bidderInfos[parentBidderName]; !ok {
>     //return err
> }
> ```

@gargcreation1992 , refer this https://github.com/prebid/prebid-server/pull/3082#discussion_r1319549765 for usage of  `parentHasSyncerCfg`  and how it is not same as `bidderInfos`.

https://github.com/prebid/prebid-server/blob/eb836cd4ae49b985afee36a0bd030fddd6e3edc7/usersync/syncersbuilder.go#L172-L173

Also no need to have separate `!ok or !parentHasSyncerCfg`  check . Line 173 makes use of `parentHasSyncerCfg` to checks whether parent has syncer config.